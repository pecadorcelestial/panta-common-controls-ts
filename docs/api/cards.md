# Card

This component wraps its content in a card-like ``div`` with a ``shadow-box`` and ``border`` of 5px.

### Properties

| Name       | Type   | Optional | Comments |
| -----------|--------|----------|----------|
| elevation  | number | true     | Sets the size of the shadow in the card. If no value is passed the box-shadow will be unset. |
| width      | string | true     | Sets the width of the card. If no value is passed the width is set to ``auto``. |

## Usage

```javascript
import React, { Component } from 'react';
import { Card } from 'panda-common-controls';

class BasicForm extends Component {
    render() {
        //Result.
        return(
            <div>
                <Card elevation={14} width='100px'>
                    {/* Put content here... */}
                </Card>
            <div>
        );
    }
}
```
