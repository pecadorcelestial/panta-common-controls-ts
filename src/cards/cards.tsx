//Componentes generales.
import * as React from 'react';
import styled, { StyledComponent } from 'styled-components';

type Elevation = 4 | 8 | 10 | 14 | 18 | 32 | undefined;

const getCardBoxShadow = (elevation: Elevation): string => {
    let result: string = '';
	switch(elevation) {
		case 4:
            result = `box-shadow:  0px 0px 4px 1px rgba(0, 0, 0, 0.05);`;
            break;
		case 8:
			result = `box-shadow:  0px 0px 8px 1px rgba(0, 0, 0, 0.05);`;
            break;
		case 10:
			result = `box-shadow:  0px 0px 10px 1px rgba(0, 0, 0, 0.1);`;
            break;
		case 14:
			result = `box-shadow:  0px 0px 14px 1px rgba(0, 0, 0, 0.1);`;
            break;
		case 18:
			result = `box-shadow:  0px 0px 18px 1px rgba(0, 0, 0, 0.1);`;
            break;
		case 32:
		default:
			result = `box-shadow:  0px 0px 32px 1px rgba(0, 0, 0, 0.1);`;
            break;
    }
    return result;
};

export interface ICardProps extends React.HTMLAttributes<HTMLDivElement> {
    elevation: Elevation;
    width?: string;
};
export const Card: StyledComponent<'div', any, ICardProps, never> = styled.div`
    background-color: #FFF;
    border-radius: 5px;
    box-sizing: border-box;
    height: auto;
    margin: 0px;
    overflow: hidden;
    padding: 0px;
    width: ${(props: ICardProps) => props.width ? props.width : `auto`};

    transition: all .3s;
        
    ${(props: ICardProps) => getCardBoxShadow(props.elevation)}		

    @media screen and (max-width: 767px) {
        border-radius: 0px;
        box-shadow: unset;
    }

    @media screen and (min-width: 768px) and (max-width: 991px) {
        border-radius: 0px;
        box-shadow: unset;
    }
`;