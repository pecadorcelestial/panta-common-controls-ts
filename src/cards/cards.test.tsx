
//RRRR  EEEEE  AAA   CCCC TTTTT
//R   R E     A   A C       T
//RRRR  EEE   AAAAA C       T
//R   R E     A   A C       T
//R   R EEEEE A   A  CCCC   T

import React from 'react';
import renderer from 'react-test-renderer';

//EEEEE N   N ZZZZZ Y   Y M   M EEEEE
//E     NN  N    Z   Y Y  MM MM E
//EEE   N N N   Z     Y   M M M EEE
//E     N  NN  Z      Y   M   M E
//EEEEE N   N ZZZZZ  YYY  M   M EEEEE

import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

enzyme.configure({ adapter: new Adapter() });

//JJJJJ EEEEE  SSSS TTTTT
//    J E     S       T
//    J EEE    SSS    T
//J   J E         S   T
// JJJ  EEEEE SSSS    T

//NOTA: Éste módulo ayuda a que los errores en componentes estilizados se vean de manera correcta:
//https://medium.com/styled-components/effective-testing-for-styled-components-67982eb7d42b
import 'jest-enzyme';
import 'jest-styled-components';

// CCCC  OOO  M   M PPPP   OOO  N   N EEEEE N   N TTTTT EEEEE  SSSS
//C     O   O MM MM P   P O   O NN  N E     NN  N   T   E     S
//C     O   O M M M PPPP  O   O N N N EEE   N N N   T   EEE    SSS
//C     O   O M   M P     O   O N  NN E     N  NN   T   E         S
// CCCC  OOO  M   M P      OOO  N   N EEEEE N   N   T   EEEEE SSSS

import { Card } from './cards';

//Snapshots.
describe('[SNAPSHOT][Componentes][Common][Card] - Tarjeta con 7 tipos diferentes de elevacion en la sombra.', () => {
	//Elevación: 4.
	it('Debe pintar el componente con una elevación en la sombra de 4.', () => {
        //Snapshot [4].
        let component = renderer.create(<Card elevation={4} width='100%'/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Elevación: 8.
	it('Debe pintar el componente con una elevación en la sombra de 8.', () => {
        //Snapshot [8].
        let component = renderer.create(<Card elevation={8}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Elevación: 10.
	it('Debe pintar el componente con una elevación en la sombra de 10.', () => {
        //Snapshot [10].
        let component = renderer.create(<Card elevation={10}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Elevación: 14.
	it('Debe pintar el componente con una elevación en la sombra de 14.', () => {
        //Snapshot [14].
        let component = renderer.create(<Card elevation={14}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Elevación: 18.
	it('Debe pintar el componente con una elevación en la sombra de 18.', () => {
        //Snapshot [18].
        let component = renderer.create(<Card elevation={18}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Elevación: 32.
	it('Debe pintar el componente con una elevación en la sombra de 32.', () => {
        //Snapshot [32].
        let component = renderer.create(<Card elevation={32}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //Elevación: N/A.
	it('Debe pintar el componente sin elevación en la sombra.', () => {
        //Snapshot [N/A].
        let component = renderer.create(<Card/>).toJSON();
		expect(component).toMatchSnapshot();
    });
});
