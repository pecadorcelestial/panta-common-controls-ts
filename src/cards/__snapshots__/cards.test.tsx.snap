// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SNAPSHOT][Componentes][Common][Card] - Tarjeta con 7 tipos diferentes de elevacion en la sombra. Debe pintar el componente con una elevación en la sombra de 4. 1`] = `
.c0 {
  background-color: #FFF;
  border-radius: 5px;
  box-sizing: border-box;
  height: auto;
  margin: 0px;
  overflow: hidden;
  padding: 0px;
  width: 100%;
  -webkit-transition: all .3s;
  transition: all .3s;
  box-shadow: 0px 0px 4px 1px rgba(0,0,0,0.05);
}

@media screen and (max-width:767px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

@media screen and (min-width:768px) and (max-width:991px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

<div
  className="c0"
  elevation={4}
  width="100%"
/>
`;

exports[`[SNAPSHOT][Componentes][Common][Card] - Tarjeta con 7 tipos diferentes de elevacion en la sombra. Debe pintar el componente con una elevación en la sombra de 8. 1`] = `
.c0 {
  background-color: #FFF;
  border-radius: 5px;
  box-sizing: border-box;
  height: auto;
  margin: 0px;
  overflow: hidden;
  padding: 0px;
  width: auto;
  -webkit-transition: all .3s;
  transition: all .3s;
  box-shadow: 0px 0px 8px 1px rgba(0,0,0,0.05);
}

@media screen and (max-width:767px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

@media screen and (min-width:768px) and (max-width:991px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

<div
  className="c0"
  elevation={8}
/>
`;

exports[`[SNAPSHOT][Componentes][Common][Card] - Tarjeta con 7 tipos diferentes de elevacion en la sombra. Debe pintar el componente con una elevación en la sombra de 10. 1`] = `
.c0 {
  background-color: #FFF;
  border-radius: 5px;
  box-sizing: border-box;
  height: auto;
  margin: 0px;
  overflow: hidden;
  padding: 0px;
  width: auto;
  -webkit-transition: all .3s;
  transition: all .3s;
  box-shadow: 0px 0px 10px 1px rgba(0,0,0,0.1);
}

@media screen and (max-width:767px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

@media screen and (min-width:768px) and (max-width:991px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

<div
  className="c0"
  elevation={10}
/>
`;

exports[`[SNAPSHOT][Componentes][Common][Card] - Tarjeta con 7 tipos diferentes de elevacion en la sombra. Debe pintar el componente con una elevación en la sombra de 14. 1`] = `
.c0 {
  background-color: #FFF;
  border-radius: 5px;
  box-sizing: border-box;
  height: auto;
  margin: 0px;
  overflow: hidden;
  padding: 0px;
  width: auto;
  -webkit-transition: all .3s;
  transition: all .3s;
  box-shadow: 0px 0px 14px 1px rgba(0,0,0,0.1);
}

@media screen and (max-width:767px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

@media screen and (min-width:768px) and (max-width:991px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

<div
  className="c0"
  elevation={14}
/>
`;

exports[`[SNAPSHOT][Componentes][Common][Card] - Tarjeta con 7 tipos diferentes de elevacion en la sombra. Debe pintar el componente con una elevación en la sombra de 18. 1`] = `
.c0 {
  background-color: #FFF;
  border-radius: 5px;
  box-sizing: border-box;
  height: auto;
  margin: 0px;
  overflow: hidden;
  padding: 0px;
  width: auto;
  -webkit-transition: all .3s;
  transition: all .3s;
  box-shadow: 0px 0px 18px 1px rgba(0,0,0,0.1);
}

@media screen and (max-width:767px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

@media screen and (min-width:768px) and (max-width:991px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

<div
  className="c0"
  elevation={18}
/>
`;

exports[`[SNAPSHOT][Componentes][Common][Card] - Tarjeta con 7 tipos diferentes de elevacion en la sombra. Debe pintar el componente con una elevación en la sombra de 32. 1`] = `
.c0 {
  background-color: #FFF;
  border-radius: 5px;
  box-sizing: border-box;
  height: auto;
  margin: 0px;
  overflow: hidden;
  padding: 0px;
  width: auto;
  -webkit-transition: all .3s;
  transition: all .3s;
  box-shadow: 0px 0px 32px 1px rgba(0,0,0,0.1);
}

@media screen and (max-width:767px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

@media screen and (min-width:768px) and (max-width:991px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

<div
  className="c0"
  elevation={32}
/>
`;

exports[`[SNAPSHOT][Componentes][Common][Card] - Tarjeta con 7 tipos diferentes de elevacion en la sombra. Debe pintar el componente sin elevación en la sombra. 1`] = `
.c0 {
  background-color: #FFF;
  border-radius: 5px;
  box-sizing: border-box;
  height: auto;
  margin: 0px;
  overflow: hidden;
  padding: 0px;
  width: auto;
  -webkit-transition: all .3s;
  transition: all .3s;
  box-shadow: 0px 0px 32px 1px rgba(0,0,0,0.1);
}

@media screen and (max-width:767px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

@media screen and (min-width:768px) and (max-width:991px) {
  .c0 {
    border-radius: 0px;
    box-shadow: unset;
  }
}

<div
  className="c0"
/>
`;
